### PR TITLE
Dev/400: patch testcase IO

### PR DESCRIPTION
**Pull request by Johnny850807**
resolve: #400
Dev/400: patch testcase IO

### CODE REVIEW

File: Commons/src/main/java/tw/waterball/judgegirl/commons/utils/ZipUtils.java [hunk 1]
```diff
@@ -156,10 +156,14 @@ public static void zipFromFile(File file, OutputStream out, String... ignoredFil
      * ignoredFileNames.
      */
     public static void zipFromFile(File[] files, OutputStream out, String... ignoredFileNames) {
+        zipFromFile(files, out, name -> ArrayUtils.contains(ignoredFileNames, name));
+    }
+
+    public static void zipFromFile(File[] files, OutputStream out, Predicate<String> ignoredFilePredicate) {
```
Comment (Johnny850807) [inline Commons/src/main/java/tw/waterball/judgegirl/commons/utils/ZipUtils.java:162]: Make it more useful and support more cases.

**Review by Chaoyuuu — APPROVED** [review]
LGTM

File: Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/repositories/ProblemRepository.java [hunk 2]
```diff
@@ -64,8 +65,39 @@
...
+            this.deletedIns = deletedIns;
+            this.deletedOuts = deletedOuts;
+            this.stdIn = stdIn;
+            this.stdOut = stdOut;
+            this.inputFiles = inputFiles;
+            this.outputFiles = outputFiles;
+        }
+
+        public Optional<FileResource> getStdIn() {
+            return Optional.ofNullable(stdIn);
```
Comment (Wally5077) [inline Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/repositories/ProblemRepository.java:95]: static import ```ofNullable```

File: Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/repositories/ProblemRepository.java [hunk 3]
```diff
@@ -64,8 +65,39 @@
...
+        }
+
+        public Optional<FileResource> getStdOut() {
+            return Optional.ofNullable(stdOut);
```
Comment (Wally5077) [inline Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/repositories/ProblemRepository.java:99]: ditto

File: Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemController.java [hunk 4]
```diff
@@ -194,22 +201,36 @@ public String uploadProvidedCodes(@RequestHeader("Authorization") String authori
...
         });
     }
+    
+    private ProblemRepository.TestcaseIoPatching testcaseIoPatching(String testcaseId, MultipartFile stdIn, MultipartFile stdOut, MultipartFile[] inFiles, MultipartFile[] outFiles, String[] deletedIns, String[] deletedOuts) {
+        return new ProblemRepository.TestcaseIoPatching(
+                testcaseId, deletedIns, deletedOuts,
+                stdIn == null ? null : convertMultipartFileToFileResource(stdIn),
+                stdOut == null ? null : convertMultipartFileToFileResource(stdOut),
+                new HashSet<>(convertMultipartFilesToFileResources(inFiles)),
+                new HashSet<>(convertMultipartFilesToFileResources(outFiles))
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemController.java:231]:
```Java
Set.copyOf(convertMultipartFilesToFileResources(inFiles));
Set.copyOf(convertMultipartFilesToFileResources(outFiles));
```

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 5]
```diff
@@ -872,11 +912,11 @@ private void problemProvidedCodesAndTestcaseIOsShouldBeDeleted(Problem problem)
...
     private void filesShouldBeSaved(String message, List<String> fileIds) {
         for (String fileId : fileIds) {
-            assertTrue(existsFile(fileId), message);
+            assertTrue(fileShouldExist(fileId), message);
         }
     }
 
-    private boolean existsFile(String fileId) {
+    private boolean fileShouldExist(String fileId) {
         return ofNullable(gridFsTemplate.findOne(new Query(where("_id").is(fileId))))
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:920]:
```Java
return ofNullable(gridFsTemplate.findOne(query(where("_id").is(fileId))))
```

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

File: Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/usecases/DownloadTestCaseIOsUseCase.java [hunk 6]
```diff
@@ -35,7 +32,9 @@ public DownloadTestCaseIOsUseCase(ProblemRepository problemRepository) {
 
     public FileResource execute(Request request) throws NotFoundException {
         return problemRepository.downloadTestCaseIOs(request.problemId, request.testcaseId)
-                .orElseThrow(() -> notFound(Problem.class).id(request.problemId));
+                .orElseThrow(() -> new NotFoundException(String.format("The testcase's IO is not found. (problemId=%d, testcaseId=%s) " +
```
Comment (Wally5077) [inline Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/usecases/DownloadTestCaseIOsUseCase.java:35]: static import ```String.format()```

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 7]
```diff
@@ -872,12 +960,12 @@ private void problemProvidedCodesAndTestcaseIOsShouldBeDeleted(Problem problem)
 
     private void filesShouldBeSaved(String message, List<String> fileIds) {
         for (String fileId : fileIds) {
-            assertTrue(existsFile(fileId), message);
+            assertTrue(fileShouldExist(fileId), message);
         }
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:964]: ```fileIds..forEach(fileId -> assertTrue(fileShouldExist(fileId), message));```

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

File: Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/usecases/DownloadTestCaseIOsUseCase.java (hunk 6 above)
Comment (Johnny850807) [reply Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/usecases/DownloadTestCaseIOsUseCase.java:35]: Needn't

**Review by Wally5077 — APPROVED** [review]
LGTM

**Pull request merged**